piControlMain.c

```diff
@@ -844,26 +844,44 @@ static long piControlIoctl(struct file *file, unsigned int prg_nr, unsigned long
 
 	case KB_GET_VALUE:
 		{
-			SPIValue *pValue = (SPIValue *) usr_addr;
+			SPIValue spi_val;
+			u8 val;
 
 			if (!isRunning())
 				return -EFAULT;
 
-			if (pValue->i16uAddress >= KB_PI_LEN) {
+			if (!access_ok(VERIFY_WRITE, (void __user *) usr_addr,
+				       sizeof(spi_val))) {
+				pr_err("invalid address provided by user\n");
+				return -EFAULT;
+			}
+
+			if (__copy_from_user(&spi_val, (const void __user *) usr_addr,
```

Comment:
Pardon my ignorance, why is the `__copy_from_user()` variant used here? AFAIU the difference between it and `copy_from_user()` is that the latter zero-fills the destination in case of a short copy. But can that ever happen given that struct SPIValueStr is apparently exactly 4 bytes, i.e. the native 32-bit CPU size?

Also, why the `access_ok()` check further up? Again, sorry if that's a dumb question.

Reply:
Also, s/an/and/ in the commit message.

Reply:
This is only a slight optimization: copy_[from|to]_user both check the userspace address by means of access_ok() before it is accessed (the check can be done for both read and write access). The __ variants of the copy functions omit this check. So we can save one call to access_ok() by doing this call upfront and call the "lightweight" versions of the copy functions.

Reply:
Admittedly it might be a bit "over the top" in this case...

Reply:
Cool, thanks for the explanation.